### PR TITLE
Magnus/gui

### PR DESCRIPTION
**Pull request by mglyngberg**
Added shuffle and loop buttons.

### CODE REVIEW

File: BusinessLayer/SongControl/SongControl.cpp [hunk 1]
```diff
@@ -16,24 +20,42 @@ SongControl::~SongControl()
...
 QString SongControl::currentSong()
 {
+    playOrder_.prepend(currentSongIndex_);
+    playOrderIndex_ = 0;
     return (files_[(currentSongIndex_)]);
 }
 
+QString SongControl::shuffleSong()
+{
+    currentSongIndex_ = (rand() % files_.size());
```
Comment (Ben-Graham) [inline BusinessLayer/SongControl/SongControl.cpp]: might want to add a check that the number generated is not the number currently playing. During a quick test of the program I had the same song play 3 times in a row in a test library of about 35 songs.

File: BusinessLayer/SongControl/SongControl.cpp [hunk 2]
```diff
@@ -16,24 +20,42 @@ SongControl::~SongControl()
...
-    if (currentSongIndex_ < 0)
+
+    if (playOrderIndex_ < 9)
     {
-        currentSongIndex_ = files_.size() - 1;
+        playOrderIndex_ += 1;
     }
-    return files_[currentSongIndex_];
+
+    if (playOrder_[playOrderIndex_] == -1)
```
Comment (Ben-Graham) [inline BusinessLayer/SongControl/SongControl.cpp]: As mentioned in person, lets have it so shuffling and going to a previous song lets plays another random song so we can keep reliable behaviour when not shuffling

**Review by Ben-Graham — CHANGES_REQUESTED** [review]

**Review by Ben-Graham — APPROVED** [review]

**Comment by Ben-Graham** [comment]
Need to fix shuffle play with only one song

File: BusinessLayer/SongControl/SongControl.cpp (hunk 1 above)
Comment (Ben-Graham) [inline BusinessLayer/SongControl/SongControl.cpp]: check for case where there is only one song in the library

**Review by Ben-Graham — CHANGES_REQUESTED** [review]

File: BusinessLayer/SongControl/SongControl.cpp [hunk 3]
```diff
@@ -34,6 +36,22 @@ QString SongControl::currentSong()
...
+QString SongControl::shuffleSong()
+{
+
+    if(files_.length() == 1)
+    {
+        return (files_[currentSongIndex_]);
+    }
+
+    int index = currentSongIndex_;
+    while(currentSongIndex_ == index)
```
Comment (bill-luu) [inline BusinessLayer/SongControl/SongControl.cpp]: Use a do while loop instead, so that the calculation is done first before checking the condition!

**Review by bill-luu — CHANGES_REQUESTED** [review]
I know you didn't do it, but I'd also rename the `SongPlayer::playFile()` function, since it just sets the media.

**Review by bill-luu — APPROVED** [review]

File: BusinessLayer/SongControl/SongControl.cpp [hunk 4]
```diff
@@ -7,6 +7,8 @@ SongControl::SongControl()
     QString dir = QString(".");
     readSongNames(dir, files_);
     currentSongIndex_ = 0;
+    shuffleSeed_ = (int)startTime_.currentTime().msec();
+    srand(shuffleSeed_);
```
Comment (Ben-Graham) [inline BusinessLayer/SongControl/SongControl.cpp]: We can just use srand(time(0)) to create random numbers. It will be simpler.

**Review by Ben-Graham — CHANGES_REQUESTED** [review]

**Review by Ben-Graham — APPROVED** [review]

**Pull request merged**